### PR TITLE
Only query the appconfig once per request

### PR DESCRIPTION
**Pull request by nickvergessen**
Problem with the old AppConfig:
- We delay the loading of each app, until an value for it is requested.
- In the beginning of each request, before loading an app, we check the type and enabled status for the app

So effectively we load all apps already in the beginning, causing `count($apps)` number of queries instead of one.

@MorrisJobke

### CODE REVIEW

**Comment by karlitschek** [comment]
interesting idea. looks good :+1:  some performance tests and results would be nice

File: lib/private/appconfig.php [hunk 1]
```diff
@@ -193,29 +159,28 @@ public function setValue($app, $key, $value) {
...
 				'configkey',
 			]);
-		}
 
-		if (!$inserted) {
-			$oldValue = $this->getValue($app, $key);
-			if($oldValue === strval($value)) {
-				return;
+			if ($inserted) {
+				$this->cache[$app][$key] = $value;
```
Comment (icewind1991) [inline lib/private/appconfig.php]: You'll need to make sure `$this->cache[$app]` exists

File: lib/private/appconfig.php [hunk 2]
```diff
@@ -245,38 +213,61 @@ public function deleteKey($app, $key) {
...
+
+		$this->cache = [];
+
+		$sql = $this->conn->getQueryBuilder();
+		$sql->select('*')
+			->from('appconfig');
+		$result = $sql->execute();
+
+		while ($row = $result->fetch()) {
+			$this->cache[$row['appid']][$row['configkey']] = $row['configvalue'];
```
Comment (icewind1991) [inline lib/private/appconfig.php]: will emit a warning for the first row in every app since `$this->cache[$row['appid']]` will not be set to an array

**Comment by icewind1991** [comment]
Looks good otherwise

File: lib/private/appconfig.php (hunk 1 above)
Comment (nickvergessen) [inline lib/private/appconfig.php]: Yeah right

**Comment by nickvergessen** [comment]
> some performance tests and results would be nice

Are non existing queries in thee query log good enough?

**Comment by icewind1991** [comment]
:+1: code looks good

**Comment by MorrisJobke** [comment]
Tested and works fine :+1:

**Comment by LukasReschke** [comment]
This probably broke Oracle? Ref https://ci.owncloud.org/job/core-ci-linux/database=oci,label=SLAVE/400/console

```
15:48:07 exception 'Doctrine\DBAL\Driver\OCI8\OCI8Exception' with message 'ORA-00932: inconsistent datatypes: expected - got CLOB' in /var/jenkins/workspace/core-ci-linux/database/oci/label/SLAVE/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Driver/OCI8/OCI8Exception.php:33
15:48:07 Stack trace:
15:48:07 #0 /var/jenkins/workspace/core-ci-linux/database/oci/label/SLAVE/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Driver/OCI8/OCI8Statement.php(214): Doctrine\DBAL\Driver\OCI8\OCI8Exception::fromErrorInfo(Array)
15:48:07 #1 /var/jenkins/workspace/core-ci-linux/database/oci/label/SLAVE/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Connection.php(987): Doctrine\DBAL\Driver\OCI8\OCI8Statement->execute()
15:48:07 #2 /var/jenkins/workspace/core-ci-linux/database/oci/label/SLAVE/lib/private/db/connection.php(203): Doctrine\DBAL\Connection->executeUpdate('UPDATE "oc_appc...', Array, Array)
15:48:07 #3 /var/jenkins/workspace/core-ci-linux/database/oci/label/SLAVE/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Query/QueryBuilder.php(208): OC\DB\Connection->executeUpdate('UPDATE `*PREFIX...', Array, Array)
15:48:07 #4 /var/jenkins/workspace/core-ci-linux/database/oci/label/SLAVE/lib/private/db/querybuilder/querybuilder.php(121): Doctrine\DBAL\Query\QueryBuilder->execute()
15:48:07 #5 /var/jenkins/workspace/core-ci-linux/database/oci/label/SLAVE/lib/private/appconfig.php(183): OC\DB\QueryBuilder\QueryBuilder->execute()
15:48:07 #6 /var/jenkins/workspace/core-ci-linux/database/oci/label/SLAVE/lib/private/app/appmanager.php(170): OC\AppConfig->setValue('files_sharing', 'enabled', 'yes')
15:48:07 #7 /var/jenkins/workspace/core-ci-linux/database/oci/label/SLAVE/lib/private/app.php(312): OC\App\AppManager->enableApp('files_sharing')
15:48:07 #8 /var/jenkins/workspace/core-ci-linux/database/oci/label/SLAVE/tests/enable_all.php(13): OC_App::enable('files_sharing')
15:48:07 #9 /var/jenkins/workspace/core-ci-linux/database/oci/label/SLAVE/tests/enable_all.php(19): enableApp('files_sharing')
15:48:07 #10 {main}
15:48:07 
15:48:07 Next exception 'Doctrine\DBAL\Exception\DriverException' with message 'An exception occurred while executing 'UPDATE "oc_appconfig" SET "configvalue" = ? WHERE ("appid" = ?) AND ("configkey" = ?) AND ("configvalue" <> ?)' with params ["yes", "files_sharing", "enabled", "yes"]:
```

cc @DeepDiver1975

**Comment by LukasReschke** [comment]
Tracked at https://github.com/owncloud/core/issues/18949

**Pull request merged**